x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/crud/rules_management_client.ts

```diff
@@ -0,0 +1,433 @@
+  ml?: SharedServices
+): IRulesManagementClient => {
+  const mlAuthz = buildMlAuthz({
+    license: licensing.license,
+    ml,
+    request,
+    savedObjectsClient,
+  });
+
+  const client = {
+    createCustomRule: async (
+      createCustomRulePayload: CreateCustomRuleProps
+    ): Promise<RuleAlertType> => {
+      return createCustomRule(rulesClient, createCustomRulePayload, mlAuthz);
+    },
+
+    createPrebuiltRule: async (
+      createPrebuiltRulePayload: CreatePrebuiltRuleProps
+    ): Promise<RuleAlertType> => {
+      return createPrebuiltRule(rulesClient, createPrebuiltRulePayload, mlAuthz);
+    },
+
+    updateRule: async (updateRulePayload: UpdateRuleProps): Promise<RuleAlertType> => {
+      return updateRule(rulesClient, updateRulePayload, mlAuthz);
+    },
+
+    patchRule: async (patchRulePayload: PatchRuleProps): Promise<RuleAlertType> => {
+      return patchRule(rulesClient, patchRulePayload, mlAuthz);
+    },
+
+    deleteRule: async (deleteRulePayload: DeleteRuleProps): Promise<void> => {
+      return deleteRule(rulesClient, deleteRulePayload);
+    },
+
+    upgradePrebuiltRule: async (
+      upgradePrebuiltRulePayload: UpgradePrebuiltRuleProps
+    ): Promise<RuleAlertType> => {
+      return upgradePrebuiltRule(rulesClient, upgradePrebuiltRulePayload, mlAuthz);
+    },
+
+    importRule: async (importRulePayload: ImportRuleProps): Promise<RuleAlertType> => {
+      return importRule(rulesClient, importRulePayload, mlAuthz);
+    },
+  };
+
+  return client;
+};
+
+export const createCustomRule = async (
+  rulesClient: RulesClient,
+  createCustomRulePayload: CreateCustomRuleProps,
+  mlAuthz: MlAuthz
+): Promise<RuleAlertType> => {
+  const { params } = createCustomRulePayload;
+  await _validateMlAuth(mlAuthz, params.type);
+
+  const rule = await _createRule(rulesClient, params, { immutable: false });
+  return rule;
+};
+
+export const createPrebuiltRule = async (
+  rulesClient: RulesClient,
+  createPrebuiltRulePayload: CreatePrebuiltRuleProps,
+  mlAuthz: MlAuthz
+): Promise<RuleAlertType> => {
+  const { ruleAsset } = createPrebuiltRulePayload;
+
+  await _validateMlAuth(mlAuthz, ruleAsset.type);
+
+  const rule = await _createRule(rulesClient, ruleAsset, {
+    immutable: true,
+    defaultEnabled: false,
+  });
+
+  return rule;
+};
+
+export const updateRule = async (
+  rulesClient: RulesClient,
+  updateRulePayload: UpdateRuleProps,
+  mlAuthz: MlAuthz
+): Promise<RuleAlertType> => {
+  const { ruleUpdate, ruleId, id } = updateRulePayload;
+
+  await _validateMlAuth(mlAuthz, ruleUpdate.type);
+
+  const existingRule = await readRules({
+    rulesClient,
+    ruleId,
+    id,
+  });
+
+  if (existingRule == null) {
+    const error = getIdError({ id, ruleId });
+    throw new ClientError(error.message, error.statusCode);
+  }
+
+  const update = await _updateRule(rulesClient, { ruleUpdate, existingRule });
+
+  const enabled = ruleUpdate.enabled ?? existingRule.enabled;
```

Comment:
Looks like logic to determine whether we need to enable or not is split between this line and `_toggleRuleEnabledOnUpdate`, which is a bit hard to read. 

Perhaps we can simplify it like:
```typescript
await _toggleRuleEnabledOnUpdate(rulesClient, existingRule, ruleUpdate.enable);
``` 
and then in `_toggleRuleEnabledOnUpdate`:
```typescript
const _toggleRuleEnabledOnUpdate = async (
  rulesClient: RulesClient,
  existingRule: RuleAlertType,
  updatedRuleEnabled: boolean
) => {
  if (existingRule.enabled && !updatedRuleEnabled) {
    await rulesClient.disable({ id: existingRule.id });
  } else if (!existingRule.enabled && updatedRuleEnabled) {
    await rulesClient.enable({ id: existingRule.id });
  }
};
```

As I understand the end result would be the same.

Reply:
Thanks, applied 👍